src/core/Table/hooks/useColumnVisibility.tsx

```diff
@@ -0,0 +1,72 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { ColumnInstance, HeaderProps, Hooks } from 'react-table';
+import { Checkbox } from '../..';
+import SvgColumnManager from '@itwin/itwinui-icons-react/cjs/icons/ColumnManager';
+import { DropdownMenu } from '../../DropdownMenu';
+import { IconButton } from '../../Buttons/IconButton';
+import { MenuItem } from '../../Menu';
+
+export const COLUMN_MANAGER_ID = 'iui-table-column-manager';
+
+export const useColumnVisibility = <T extends Record<string, unknown>>(
+  hasColumnManager: boolean,
+) => (hooks: Hooks<T>) => {
+  if (!hasColumnManager) {
+    return;
+  }
+
+  hooks.allColumns.push((columns: ColumnInstance<T>[]) => [
+    ...columns,
+    {
+      id: COLUMN_MANAGER_ID,
+      minWidth: 48,
+      width: 48,
+      maxWidth: 48,
+      disableReordering: true,
+      columnClassName: 'iui-slot',
+      cellClassName: 'iui-slot',
+      Header: ({ allColumns }: HeaderProps<T>) => {
+        const [isOpen, setIsOpen] = React.useState(false);
+        const headerCheckBoxes = () =>
+          allColumns
+            .filter(({ id }) => !id.includes('iui-table'))
+            .map((column) => {
+              const { onChange, checked } = column.getToggleHiddenProps();
+              return (
+                <MenuItem
+                  key={column.id}
+                  icon={
+                    <Checkbox
+                      id={`iui-column-${column.id}`}
+                      checked={checked}
+                      disabled={column.disableToggleVisibility}
+                      onClick={(e) => e.stopPropagation()}
+                      onChange={onChange}
+                    />
+                  }
+                  onClick={() => onChange({ target: { checked: !checked } })}
```

Comment:
Maybe change to `() => column.toggleHidden(!checked)` as in `react-table` https://github.com/TanStack/react-table/blob/v7/src/hooks/useColumnVisibility.js#L35

Reply:
Changed. I used `() => column.toggleHidden(checked` instead of `!checked` because the `react-table` story hides columns that are checked when we are trying to show columns that are checked